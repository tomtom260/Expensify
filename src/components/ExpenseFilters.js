import React from 'react';
import 'react-dates/initialize';
import {DateRangePicker} from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';
import moment from 'moment';
import {connect} from 'react-redux'

import { addTextFilter, sortByAlphabeticalOrder, sortByAmount, 
    sortByDate,setEndDate,setStartDate, setStartAmount, setEndAmount } from '../action/filters';
    
export class ExpenseFilters extends React.Component{
    state={
        focusedInput:null
    }
    
    render(){
        return(
            <div>   
                <input
                    type="text"
                    value={this.props.filters.text}
                    onChange={e=>this.props.dispatch(addTextFilter(e.target.value))}
                />
                <select
                    value={this.props.filters.sortBy}
                    onChange={e=>{
                        switch(e.target.value){
                            default: 
                            case "DATE":return this.props.dispatch(sortByDate())
                            case "AMOUNT":return this.props.dispatch(sortByAmount())
                            case "ALPHABETICAL": return this.props.dispatch(sortByAlphabeticalOrder())
                        }
        
                    }}
                >
                    <option value="DATE">Date</option>
                    <option value="AMOUNT">Amount</option>
                    <option value="ALPHABETICAL">Alphabetical</option>
                </select>
                    <DateRangePicker
                        startDate={this.props.filters.startDate?moment(this.props.filters.startDate):null}
                        endDate={this.props.filters.endDate?moment(this.props.filters.endDate):null}
                        focusedInput={this.state.focusedInput}
                        onFocusChange={(focusedInput)=>this.setState(()=>({focusedInput}))}
                        onDatesChange={({ startDate, endDate }) =>{
                            this.props.dispatch(setStartDate(startDate))
                            this.props.dispatch(setEndDate(endDate))
                        }}
                        startDateId="your_unique_start_date_id"
                        endDateId="your_unique_end_date_id"
                        numberOfMonths={1}
                        hideKeyboardShortcutsPanel
                        isOutsideRange={()=>false}
                        showClearDates
                    />
                    <input
                        type="text"
                        value={this.props.filters.startAmount||""}
                        onChange={e=>{
                            if(e.target.value.match(/^\d*(\.)?(\d{1,2})?$/))
                                this.props.dispatch(setStartAmount(parseFloat(e.target.value,10)))
                        }}
                        placeholder="Start Amount"
                    />
                    <input
                        type="text"
                        value={this.props.filters.endAmount||""}
                        onChange={e=>{
                            if(e.target.value.match(/^\d*(\.)?(\d{1,2})?$/))
                                this.props.dispatch(setEndAmount(parseFloat(e.target.value,10)))
                        }}
                        placeholder="End Amount"
                    />
            </div>
        )
    }
}


const mapStateToProps= state =>({
    filters:state.filters
})

export default connect(mapStateToProps)(ExpenseFilters);
