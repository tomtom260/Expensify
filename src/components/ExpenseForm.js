import React from 'react';
import 'react-dates/initialize';
import {SingleDatePicker} from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';
import moment from 'moment';

class ExpenseForm extends React.Component{
    state={
        description:this.props.description||"",
        amount:this.props.amount||"",
        note:this.props.note||"",
        createdAt:moment(this.props.createdAt)||moment(Date.now()),
        buttonDisabled :this.props.history.location.pathname.includes('create'),
        showDatePicker:false
    }

    buttonDisabled = (description,amount,createdAt)=>{
        if(!description)
            return this.setState({buttonDisabled:true})
        
        if(!amount||!amount.match(/^\d+((\.)\d{1,2})?$/))
            return this.setState({buttonDisabled:true})
        
        if(!createdAt||typeof createdAt.valueOf()!=='number')
            return this.setState({buttonDisabled:true})

        return this.setState({buttonDisabled:false})
    }

    
    onSubmit = (e)=>{
        e.preventDefault()
        const expense = {
            amount:parseFloat(this.state.amount,10),
            description:this.state.description,
            createdAt:this.state.createdAt.valueOf(),
            note:this.state.note
        }
        this.props.onSubmit(expense,this.props.id)
        this.props.history.push('/');
    }


    render(){
        return(
            <form>
                <input 
                placeholder ="Description"
                autoFocus
                value={this.state.description}
                onChange={e=>{
                    const description=e.target.value;
                    this.setState({description})
                    this.buttonDisabled(description,this.state.amount,this.state.createdAt)
                }}
            />
                <input 
                placeholder ="Amount"
                value={this.state.amount}
                onChange={e=>{
                    const amount=e.target.value;
                    if(amount.match(/^\d*(\.)?(\d{1,2})?$/)){
                        this.setState({amount})
                        this.buttonDisabled(this.state.description,amount,this.state.createdAt)
                    }
                }}
                />

                <SingleDatePicker
                    date={this.state.createdAt}
                    onDateChange={createdAt => {this.setState({
                                createdAt:createdAt
                            })
                            this.buttonDisabled(this.state.description,this.state.amount,createdAt)
                        }
                    }                       
                    focused={this.state.showDatePicker}
                    onFocusChange={({focused})=>this.setState({showDatePicker:focused})}
                    numberOfMonths={1}
                    hideKeyboardShortcutsPanel
                    isOutsideRange={()=>false}
                    showClearDate
                    placeholder="DD/MM/YYYY"
                />
                <textarea 
                    placeholder="Note (Optional)"
                    value={this.state.note}
                    onChange={e=>{
                        const note=e.target.value
                        this.setState({note})
                    }}

                ></textarea>
                <button 
                    disabled={this.state.buttonDisabled}
                    onClick={this.onSubmit}
                >
                    {this.props.history.location.pathname.includes('create')?'Add':'Edit'} Expense
                </button>
            </form>
        )

    }
}


export default ExpenseForm