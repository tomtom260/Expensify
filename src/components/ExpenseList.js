import React from 'react';
import {Link} from 'react-router-dom'
import {connect} from 'react-redux'

import filterExpenses from '../selectors/filterExpenses'
import { removeExpense } from '../action/expenses';

export const ExpenseList =  props => (
    <div>
        {
            props.expenses.length===0?
                <p>No Expenses</p>:
            (props.expenses.map(el=>{
                return <div key={el.id}>
                    <Link to={`/edit/${el.id}`}>
                        <h2>{el.description}</h2>
                        <p>Amount: {el.amount}  createdAt: {el.createdAt} </p>
                    </Link>
                    <button onClick={()=>props.dispatch(removeExpense(el.id))}>Remove</button>
                </div>
            }))
        }
    </div>
)


const mapStatetoProps= state =>({
    expenses:filterExpenses(state),
})

export default connect(mapStatetoProps)(ExpenseList);