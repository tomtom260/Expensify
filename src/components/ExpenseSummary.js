import React from 'react'
import {connect} from 'react-redux';
import numeral from 'numeral';

import filterExpenses from '../selectors/filterExpenses'

const calculateTotal = (expense)=>expense.reduce((acc,cur)=>acc+cur.amount,0)

export const ExpenseSummary = props=>(
    <div>
        <h1>
            Viewing
            <span className="TotalExpense__numberOfExpenses"> {props.expenses.length} </span>
            Expense{props.expenses.length?'s':""} totalling
            <span className="TotalExpense__totalExpense"> {numeral(calculateTotal(props.expenses)).format('0,0.00')} </span>
            Birr.
        </h1>
    </div>
)

const mapStateToProps = (state)=>({
    expenses:filterExpenses(state)
})

export default connect(mapStateToProps)(ExpenseSummary);

