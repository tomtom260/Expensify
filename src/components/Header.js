import React, { Fragment } from 'react';
import { NavLink } from 'react-router-dom';

const Header =()=>(
    <Fragment>
        <h1>Expensify</h1>
        <NavLink exact activeClassName="activeLink" to="/">Home</NavLink>
        <NavLink activeClassName="activeLink" to="/create">Create Expense</NavLink>
    </Fragment>
)

export default Header;