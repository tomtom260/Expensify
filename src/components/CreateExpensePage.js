import React from 'react'
import { connect } from 'react-redux';
import { addExpense } from '../action/expenses';
import ExpenseForm from './ExpenseForm';

const CreateExpensePage =(props)=>(
    <div>
        <h1>Create New Expense</h1>
        <ExpenseForm onSubmit={(expense)=>props.dispatch(addExpense(expense))} history={props.history}/>
    </div>
)

export default connect()(CreateExpensePage);