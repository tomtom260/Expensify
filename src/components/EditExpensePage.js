import React from 'react'
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {editExpense,removeExpense} from '../action/expenses'

export const EditExpensePage =(props)=>(
    <div>
        <h1>Edit Expense</h1>
        <ExpenseForm 
            onSubmit={(expense,id)=>props.dispatch(editExpense(id,expense))}
            {...props.expense} 
            history={props.history} />
        <button onClick={()=>props.dispatch(removeExpense(props.match.params.id))}>Remove Expense</button>    
    </div>
)

const mapPropsFromState = (state,props) =>{
    const expense = state.expenses.find((el)=>el.id===props.match.params.id)
    return{
        expense
    }
}
 
export default connect(mapPropsFromState)(EditExpensePage);