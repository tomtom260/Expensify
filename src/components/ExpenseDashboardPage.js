import React from 'react';
import ExpenseFilters from './ExpenseFilters'
import ExpenseList from './ExpenseList'
import ExpenseSummary from './ExpenseSummary';

export default ()=>(
    <div>
        <ExpenseSummary/>
        <ExpenseFilters />
        <ExpenseList  />  
    </div>
)
