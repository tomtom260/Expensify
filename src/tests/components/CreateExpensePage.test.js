import React from 'react';
import {shallow} from 'enzyme';
import {v4 as uuid} from 'uuid'

import expenseData from '../fixtures/expenses'
import {CreateExpensePage} from '../../components/CreateExpensePage';

test('should render createExpensePage correctly',()=>{
    const wrapper = shallow(<CreateExpensePage />)
    expect(wrapper).toMatchSnapshot()
})

test('dispatch should be called with the add expense action object',()=>{
    const expense = {
        ...expenseData[0]
    }
    delete expense.id

    const dispatchSpy = jest.fn()
    const wrapper= shallow(<CreateExpensePage dispatch={dispatchSpy}/>)
    wrapper.find('ExpenseForm').prop('onSubmit')(expense);
    expect(dispatchSpy).toHaveBeenCalledWith({
        type:"ADD_EXPENSE",
        expense:{
            description:expenseData[0].description,
            amount:expenseData[0].amount,
            createdAt:expenseData[0].createdAt,
            note:expenseData[0].note,
            id:uuid()
        }
    })
})