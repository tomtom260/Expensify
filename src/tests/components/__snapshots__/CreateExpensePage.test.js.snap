// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render createExpensePage correctly 1`] = `
<div>
  <h1>
    Create New Expense
  </h1>
  <ExpenseForm
    onSubmit={[Function]}
  />
</div>
`;
